### PR TITLE
Swift Package product type for embedding

### LINKED ISSUE #1820: Add support for dynamic Swift Package libraries
**Describe the bug**
Adding local package to the project and as dependency to iOS app target does link it, but does not embed & sign it.

**To Reproduce**
Project.swift
```swift
packages: [
        // Local
        .package(path: "./TWLogging/"),
]
```

In app target:
```swift
dependencies: [
        .package(product: "TWLogging"),
]
```

**Expected behavior**
Tuist sets up correctly embedding of the package in the app. Links, embeds and signs.

**Screenshots**
Screenshot 2020-09-24 at 16 11 06

**Desktop (please complete the following information):**
- Version 1.18.1

### PR DESCRIPTION
**Pull request by zdnk**
Resolves #1820 

### Short description 📝

Add support for linking and embedding dynamic products of Swift packages.

### Implementation 👩‍💻👨‍💻

- [x] add dynamic product to embed phase
- [ ] add Graph dependency case, transitive deps
- [ ] make test with the new dynamic option
- [ ] update documentation
- [ ] add changelog entry

### CODE REVIEW

**Comment by zdnk** [comment]
I need help with this. Not really sure how to add test for this properly as I am having trouble finding the place where the packages are tested.

Also my solution generates the project fine, the library shows as `Embed & sign` in Xcode, but when trying to install CocoaPods, its fails with 
> RuntimeError - [Xcodeproj] Consistency issue: no parent for object `BuildFile`: `Embed Frameworks`, `FrameworksBuildPhase`

I guess I've dont something wrong or theres a piece of puzzle missing.

File: fixtures/ios_app_with_local_swift_package/Packages/PackageA/Package.swift [hunk 1]
```diff
@@ -12,6 +12,7 @@ let package = Package(
             targets: ["LibraryA"]),
         .library(
             name: "LibraryB",
+            type: .dynamic,
```
Comment (kwridan) [inline fixtures/ios_app_with_local_swift_package/Packages/PackageA/Package.swift:15]: Nice, this should help us test / validate the end results 👍

**Review by kwridan — COMMENTED** [review]
Thanks for making a start on this @zdnk

File: Sources/TuistGenerator/Generator/LinkGenerator.swift [hunk 2]
```diff
@@ -465,5 +468,19 @@ extension PBXTarget {
...
+            guard let embedPhase = embedFrameworksBuildPhases().first else {
+                throw "No frameworks embed phase"
+            }
+
+            buildFile.settings = ["ATTRIBUTES": ["CodeSignOnCopy"]]
+            embedPhase.files?.append(buildFile)
+
+        case .staticLibrary:
+            break
+        }
```
[the remark this answers is not in the document]
Comment (zdnk) [reply Sources/TuistGenerator/Generator/LinkGenerator.swift:484]: > * Update `GraphDependencyReference` to include a new case for Swift packages with the appropriate properties (by the looks of it `name` and `path`).

And product type? or should there be only dynamic libraries?
Where/how would I get the path?

> Do you have push access to tuist ?

I don't think so, I tried to push the new branch directly to tuist as I was thinking the same about collab.

> Another option is we can tackle this incrementally, i.e. we start with the simple approach taken here provided it works without issues and follow up with the Graph integration to cater for the the more complex cases.

Well, this dolution currently breaks something in the project as CocoaPods is unable to install the pods as I mentioned in the comment above.

File: Sources/TuistCore/Models/PackageProductType.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import ProjectDescription
```
[the remark this answers is not in the document]
Comment (zdnk) [reply Sources/TuistCore/Models/PackageProductType.swift:1]: Thanks, I will fix it

[the remark this answers is not in the document]
Comment (zdnk) [reply Sources/TuistCore/Models/PackageProductType.swift]: Fixed.

**Comment by zdnk** [comment]
@kwridan I have updated to code according to your instructions/guide. I also gave you access to my fork so we can collaborate at least there.

I inserted `#error` in places I wasn't sure how to implement as my knowledge of this domain is still very limited.
Are there any resources I could use to get in the picture so I could be more helpful next time?

**Comment by kwridan** [comment]
> @kwridan I have updated to code according to your instructions/guide. I also gave you access to my fork so we can collaborate at least there.

Awesome, thanks! will take a look at the branch

**Comment by kwridan** [comment]
thanks for the last set of updates @zdnk they were all in the right spots! 👍 

I pushed a few additional updates + some tests .

I discovered an issue that may occur when two projects reference the same package, we currently end up with duplicate link / embed products because of the `path` used internally for the package. 

https://github.com/tuist/tuist/pull/1831/files#diff-28ff95dabe43c7ec91db99ae5cecdee8R985

**Comment by zdnk** [comment]
Hey, thanks! :)

> I discovered an issue that may occur when two projects reference the same package, we currently end up with duplicate link / embed products because of the path used internally for the package.

Not sure how to solve this one.

However the tests, are also failing because currently there are no schemes generated for local package products in the project.

**Comment by zdnk** [comment]
What would be next steps now then?

**Comment by kwridan** [comment]
Hi @zdnk, I haven't had a chance to look further unfortunately, should be able to take another pass soon.

re: the schemes, locally testing I see them, were you referring to the `ios_app_with_local_swift_package` fixture?

```sh
$ xcrun xcodebuild -list

Information about project "App":
    Targets:
        App
        AppTests

    Build Configurations:
        Debug
        Release

    If no build configuration is specified and -scheme is not passed then "Release" is used.

    Schemes:
        App
        AppTests
        LibraryA
        LibraryB
```

As for what next so the bits I could think of:

- Attempting to troubleshoot the failing cases / fixing them 
- More testing on more scenarios to asses if there are any other edge cases (e.g. we haven't tried remote packages yet)
- Adding lint rules to ensure the same package product doesn't have inconsistent product type declaration (e.g. specified as `.static` in one target but `.dynamic` in another
- Updating docs about feature and limitations (e.g. warning specifying `.dynamic` for `.static` products can cause issues or vice versa)

**Comment by kwridan** [comment]
@zdnk I've made a bit of further progress on this - apologies this is proving to be tricker than I'd had originally hoped :(

There are still some issues when multiple projects depend on the same packages that could yield duplicate linked libraries.

**Comment by ollieatkinson** [comment]
@kwridan 
> There are still some issues when multiple projects depend on the same packages that could yield duplicate linked libraries.

We are currently allowing you to specify the package type as part of the dependency requirement, which is different from how swift package manager manages these so undoubtably we will run into problems if we are trying to force another type of linking than the developer specified.

I think Tuist would probably fit better if it automatically chose the linking type depending on how that library is being used - what does it depend on, and what depends on it... some food for thought

**Comment by kwridan** [comment]
Thanks for the feedback @ollieatkinson !

Indeed auto deciding linkage would be the best option, is there a way to do that though? 

To my knowledge the linkage type is defined within the source package manifest, making it hard for Tuist to override / alter it :(

**Comment by ollieatkinson** [comment]
> To my knowledge the linkage type is defined within the source package manifest, making it hard for Tuist to override / alter it :(

If the package defines the linkage, we should not override it in Tuist - if it's auto (which is the default) we can take over and figure out what linkage is best based on it's place in the graph. 

As to how, there's a few ways we can try:
- it would be worth researching how Xcode manages `auto`
- Swift Package Manager has an old script to generate an xcodeproj, what is it doing to manage `auto`?
- Can we generate a Tuist project from a Package.swift

**Comment by pavm035** [comment]
Any updates on this we are currently blocked due to this issue?
https://github.com/tuist/tuist/issues/3425
